setup/src/Magento/Setup/Test/Unit/Module/Di/Code/Reader/FileClassScannerTest.php

```diff
@@ -0,0 +1,244 @@
+
+        $result = $scanner->getClassNames();
+        self::assertCount(0, $result);
+    }
+
+    public function testGetClassName()
+    {
+        $scanner = $this->getMockBuilder(FileClassScanner::class)->disableOriginalConstructor()->setMethods([
+            'getFileContents'
+        ])->getMock();
+        $scanner->expects(self::once())->method('getFileContents')->willReturn(
+<<<PHP
+<?php
+
+class ThisIsATest {
+
+}
+PHP
+        );
+        /* @var $scanner FileClassScanner */
+
+        $result = $scanner->getClassNames();
+
+        self::assertCount(1, $result);
+        self::assertContains('ThisIsATest', $result);
+    }
+
+    public function testGetClassNameAndSingleNamespace()
+    {
+        $scanner = $this->getMockBuilder(FileClassScanner::class)->disableOriginalConstructor()->setMethods([
+            'getFileContents'
+        ])->getMock();
+        $scanner->expects(self::once())->method('getFileContents')->willReturn(
+<<<PHP
+<?php
+
+namespace NS;
+
+class ThisIsMyTest {
+
+}
+PHP
+        );
+        /* @var $scanner FileClassScanner */
+
+        $result = $scanner->getClassNames();
+
+        self::assertCount(1, $result);
+        self::assertContains('NS\ThisIsMyTest', $result);
+    }
+
+    public function testGetClassNameAndMultiNamespace()
+    {
+        $scanner = $this->getMockBuilder(FileClassScanner::class)->disableOriginalConstructor()->setMethods([
+            'getFileContents'
+        ])->getMock();
+        $scanner->expects(self::once())->method('getFileContents')->willReturn(
+<<<PHP
+<?php
+
+namespace This\Is\My\Ns;
+
+class ThisIsMyTest {
+
+    public function __construct()
+    {
+        \This\Is\Another\Ns::class;
+    }
+    
+    public function test()
+    {
+        
+    }
+}
+PHP
+        );
+        /* @var $scanner FileClassScanner */
+
+        $result = $scanner->getClassNames();
+
+        self::assertCount(1, $result);
+        self::assertContains('This\Is\My\Ns\ThisIsMyTest', $result);
+    }
+
+
+    public function testGetMultiClassNameAndMultiNamespace()
+    {
+        $scanner = $this->getMockBuilder(FileClassScanner::class)->disableOriginalConstructor()->setMethods([
+            'getFileContents'
+        ])->getMock();
+        $scanner->expects(self::once())->method('getFileContents')->willReturn(
+<<<PHP
+<?php
+
+namespace This\Is\My\Ns;
+
+class ThisIsMyTest {
+
+    public function __construct()
+    {
+        \$this->get(\This\Is\Another\Ns::class)->method();
+        self:: class;
+    }
+    
+    public function test()
+    {
+        
+    }
+}
+
+class ThisIsForBreaking {
+
+}
+
+PHP
+        );
+        /* @var $scanner FileClassScanner */
+
+        $result = $scanner->getClassNames();
+
+        self::assertCount(2, $result);
+        self::assertContains('This\Is\My\Ns\ThisIsMyTest', $result);
+        self::assertContains('This\Is\My\Ns\ThisIsForBreaking', $result);
+    }
+
+    public function testBracketedNamespacesAndClasses()
```

Comment:
@kschroeder what do you think about extracting classes to data provider(s) here? It would be more flexible to add more cases in future. I think you could make these tests cleaner this way.